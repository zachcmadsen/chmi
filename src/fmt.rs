use tracing::{Level, Subscriber};
use tracing_subscriber::{
    fmt::{self, format, FormatEvent, FormatFields},
    registry::LookupSpan,
};

pub struct Formatter(());

impl Formatter {
    pub fn new() -> Formatter {
        Formatter(())
    }
}

impl<S, N> FormatEvent<S, N> for Formatter
where
    S: Subscriber + for<'a> LookupSpan<'a>,
    N: for<'a> FormatFields<'a> + 'static,
{
    fn format_event(
        &self,
        ctx: &fmt::FmtContext<'_, S, N>,
        mut writer: format::Writer<'_>,
        event: &tracing::Event<'_>,
    ) -> std::fmt::Result {
        write!(writer, "chmi: ")?;

        let level = match *event.metadata().level() {
            Level::TRACE => "trace",
            Level::DEBUG => "debug",
            Level::INFO => "info",
            Level::WARN => "warn",
            Level::ERROR => "error",
        };
        write!(writer, "{}: ", level)?;

        ctx.format_fields(writer.by_ref(), event)?;

        writeln!(writer)
    }
}
