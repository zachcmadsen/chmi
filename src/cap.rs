use std::fmt;

pub const INPUT_SELECT_CODE: u8 = 0x60;

#[derive(Debug, PartialEq)]
pub struct VcpCode {
    pub code: u8,
    pub values: Vec<u8>,
}

#[derive(Debug)]
pub struct Capabilities {
    pub vcp: Option<Vec<VcpCode>>,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Input {
    DisplayPort1,
    DisplayPort2,
    Hdmi1,
    Hdmi2,
}

impl fmt::Display for Input {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self {
            Input::DisplayPort1 => write!(f, "DisplayPort 1"),
            Input::DisplayPort2 => write!(f, "DisplayPort 2"),
            Input::Hdmi1 => write!(f, "HDMI 1"),
            Input::Hdmi2 => write!(f, "HDMI 2"),
        }
    }
}

impl From<Input> for u8 {
    fn from(value: Input) -> Self {
        match value {
            Input::DisplayPort1 => 0x0F,
            Input::DisplayPort2 => 0x10,
            Input::Hdmi1 => 0x11,
            Input::Hdmi2 => 0x12,
        }
    }
}

impl TryFrom<u8> for Input {
    type Error = ();

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0x0F => Ok(Input::DisplayPort1),
            0x10 => Ok(Input::DisplayPort2),
            0x11 => Ok(Input::Hdmi1),
            0x12 => Ok(Input::Hdmi2),
            _ => Err(()),
        }
    }
}

impl Capabilities {
    pub fn has_input_select(&self) -> bool {
        self.vcp.as_ref().is_some_and(|vcp_codes| {
            vcp_codes.iter().any(|vcp_code| vcp_code.code == INPUT_SELECT_CODE)
        })
    }

    pub fn inputs(&self) -> Option<Vec<Input>> {
        let mut inputs = Vec::new();

        let vcp_codes = self.vcp.as_ref()?;
        let vcp_code = vcp_codes
            .iter()
            .find(|vcp_code| vcp_code.code == INPUT_SELECT_CODE)?;
        for value in &vcp_code.values {
            if let Ok(input) = (*value).try_into() {
                inputs.push(input);
            }
        }

        Some(inputs)
    }
}
