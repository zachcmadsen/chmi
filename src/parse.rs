use std::fmt;

use anyhow::{anyhow, Context};
use logos::Logos;

use crate::cap::{Capabilities, VcpCode};

#[derive(Clone, Copy, Debug, Logos, PartialEq)]
#[logos(skip "[ \x00]")]
enum Token {
    #[token("(")]
    LeftParen,
    #[token(")")]
    RightParen,

    #[token("vcp")]
    Vcp,
    #[regex("[0-9A-F][0-9A-F]", |lex| u8::from_str_radix(lex.slice(), 16).unwrap())]
    HexNumber(u8),

    #[regex("[a-zA-Z0-9_\\.]+")]
    Unknown,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match &self {
            Token::LeftParen => "'('",
            Token::RightParen => "')'",
            Token::Vcp => "'vcp'",
            Token::HexNumber(_) => "hexadecimal number",
            Token::Unknown => "unknown",
        };
        write!(f, "{}", s)
    }
}

struct CapabilitiesStringParser<'a> {
    tokens: &'a [Token],
    index: usize,
}

impl<'a> CapabilitiesStringParser<'a> {
    fn new(tokens: &'a [Token]) -> CapabilitiesStringParser<'a> {
        CapabilitiesStringParser { tokens, index: 0 }
    }

    fn parse(&mut self) -> anyhow::Result<Capabilities> {
        let mut capabilities = Capabilities { vcp: None };

        self.expect(Token::LeftParen)?;
        while !self.check(Token::RightParen) {
            match self.next()? {
                Token::Vcp => capabilities.vcp = Some(self.parse_vcp()?),
                Token::Unknown => {
                    self.expect(Token::LeftParen)?;
                    self.eat_until(Token::RightParen);
                    self.expect(Token::RightParen)?;
                }
                _ => panic!("invalid capabilities string"),
            };
        }
        self.expect(Token::RightParen)?;

        Ok(capabilities)
    }

    fn parse_vcp(&mut self) -> anyhow::Result<Vec<VcpCode>> {
        self.expect(Token::LeftParen)?;
        let mut vcp_codes = Vec::new();
        while !self.check(Token::RightParen) {
            let vcp_code = self.parse_vcp_code()?;
            vcp_codes.push(vcp_code)
        }
        self.expect(Token::RightParen)?;
        Ok(vcp_codes)
    }

    fn parse_vcp_code(&mut self) -> anyhow::Result<VcpCode> {
        let code = self.parse_number()?;
        let mut values = Vec::new();
        if self.eat(Token::LeftParen) {
            while !self.check(Token::RightParen) {
                let value = self.parse_number()?;
                values.push(value);
            }
            self.expect(Token::RightParen)?;
        };
        Ok(VcpCode { code, values })
    }

    fn parse_number(&mut self) -> anyhow::Result<u8> {
        match self.next()? {
            Token::HexNumber(n) => Ok(n),
            token => {
                return Err(anyhow!(
                    "expected hexadecimal number, found {}",
                    token
                ))
            }
        }
    }

    /// Consumes and expects `token`.
    fn expect(&mut self, token: Token) -> anyhow::Result<()> {
        let t = self.next()?;
        if t == token {
            Ok(())
        } else {
            Err(anyhow!("expected {}, found {}", token, t))
        }
    }

    /// Consumes the next token if it's `token`, and returns whether the token
    /// was consumed.
    fn eat(&mut self, token: Token) -> bool {
        self.check(token)
            .then(|| {
                self.index += 1;
            })
            .is_some()
    }

    /// Consumes tokens until the next token is `token`.
    fn eat_until(&mut self, token: Token) {
        while self.index < self.tokens.len() && !self.check(token) {
            self.index += 1;
        }
    }

    /// Returns true if the next token is `token`.
    fn check(&self, token: Token) -> bool {
        self.tokens.get(self.index).is_some_and(|&t| t == token)
    }

    /// Returns the next token.
    fn next(&mut self) -> anyhow::Result<Token> {
        self.tokens
            .get(self.index)
            .map(|t| {
                self.index += 1;
                t.clone()
            })
            .ok_or(anyhow!("unexpected end-of-file"))
    }
}

pub fn parse(capabilities_string: &str) -> anyhow::Result<Capabilities> {
    let mut tokens = Vec::new();
    for (token, span) in Token::lexer(capabilities_string).spanned() {
        match token {
            Ok(token) => tokens.push(token),
            Err(_) => {
                return Err(anyhow::anyhow!(
                    "unexpected character(s) '{}'",
                    &capabilities_string[span.start..span.end],
                ))
            }
        }
    }

    CapabilitiesStringParser::new(&tokens)
        .parse()
        .context("failed to parse capabilities string")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_un880_capabilities() {
        let capabilities_string = "(prot(monitor)type(lcd)UN880cmds(01 02 03 0C E3 F3)vcp(02 04 05 08 10 12 14(05 08 0B ) 16 18 1A 52 60( 11 12 0F 00) AC AE B2 B6 C0 C6 C8 C9 D6(01 04) DF 62 8D F4 F5(00 01 02) F6(00 01 02) 4D 4E 4F 15(01 06 11 13 14 15 18 19 28 29 48) F7(00 01 02 03) F8(00 01) F9 E4 E5 E6 E7 E8 E9 EA EB EF FD(00 01) FE(00 01 02) FF)mccs_ver(2.1)mswhql(1))";
        let capabilities = parse(&capabilities_string).unwrap();
        insta::assert_debug_snapshot!(capabilities);
    }

    #[test]
    fn parse_u32j59x_capabilities() {
        let capabilities_string = "(prot(monitor)type(lcd)SAMSUNGcmds(01 02 03 07 0C E3 F3)vcp(02 04 05 08 10 12 14(05 08 0B 0C) 16 18 1A 52 60( 11 12 0F) AC AE B2 B6 C6 C8 C9 D6(01 04 05) DC(00 02 03 05 ) DF FD)mccs_ver(2.1)mswhql(1))";
        let capabilities = parse(&capabilities_string).unwrap();
        insta::assert_debug_snapshot!(capabilities);
    }

    #[test]
    fn parse_vg259_capabilities() {
        let capabilities_string = "(prot(monitor) type(LCD)model(VG259) cmds(01 02 03 07 0C F3) vcp(02 04 05 08 10 12 14(05 06 08 0B) 16 18 1A 52 60(11 12 0F) 62 6C 6E 70 86(02 0B) 87(00 0A 14 1E 28 32 3C 46 50 5A 64) 8A 8D(01 02) AC AE B6 C6 C8 C9 CC(01 02 03 04 05 06 07 08 09 0A 0C 0D 11 12 14 1A 1E 1F 23 30 31) D6(01 05) DC(01 02 03 04 05 06 07 08) DF E0(00 01 02 03 04 05) E1(00 01) E3(00 01 02 03 04 05 06) E4(00 01 02 03 04 05) E5(00 01 02 03) E6(00 01 02 03 04) E7(00 01) E9(00 01) EA(00 01) EB(00 01))mccs_ver(2.2)asset_eep(32)mpu(01)mswhql(1))";
        let capabilities = parse(&capabilities_string).unwrap();
        insta::assert_debug_snapshot!(capabilities);
    }
}
