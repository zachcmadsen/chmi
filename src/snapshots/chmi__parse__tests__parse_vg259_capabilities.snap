---
source: src/parse.rs
expression: capabilities
---
Capabilities {
    vcp: Some(
        [
            VcpCode {
                code: 2,
                values: [],
            },
            VcpCode {
                code: 4,
                values: [],
            },
            VcpCode {
                code: 5,
                values: [],
            },
            VcpCode {
                code: 8,
                values: [],
            },
            VcpCode {
                code: 16,
                values: [],
            },
            VcpCode {
                code: 18,
                values: [],
            },
            VcpCode {
                code: 20,
                values: [
                    5,
                    6,
                    8,
                    11,
                ],
            },
            VcpCode {
                code: 22,
                values: [],
            },
            VcpCode {
                code: 24,
                values: [],
            },
            VcpCode {
                code: 26,
                values: [],
            },
            VcpCode {
                code: 82,
                values: [],
            },
            VcpCode {
                code: 96,
                values: [
                    17,
                    18,
                    15,
                ],
            },
            VcpCode {
                code: 98,
                values: [],
            },
            VcpCode {
                code: 108,
                values: [],
            },
            VcpCode {
                code: 110,
                values: [],
            },
            VcpCode {
                code: 112,
                values: [],
            },
            VcpCode {
                code: 134,
                values: [
                    2,
                    11,
                ],
            },
            VcpCode {
                code: 135,
                values: [
                    0,
                    10,
                    20,
                    30,
                    40,
                    50,
                    60,
                    70,
                    80,
                    90,
                    100,
                ],
            },
            VcpCode {
                code: 138,
                values: [],
            },
            VcpCode {
                code: 141,
                values: [
                    1,
                    2,
                ],
            },
            VcpCode {
                code: 172,
                values: [],
            },
            VcpCode {
                code: 174,
                values: [],
            },
            VcpCode {
                code: 182,
                values: [],
            },
            VcpCode {
                code: 198,
                values: [],
            },
            VcpCode {
                code: 200,
                values: [],
            },
            VcpCode {
                code: 201,
                values: [],
            },
            VcpCode {
                code: 204,
                values: [
                    1,
                    2,
                    3,
                    4,
                    5,
                    6,
                    7,
                    8,
                    9,
                    10,
                    12,
                    13,
                    17,
                    18,
                    20,
                    26,
                    30,
                    31,
                    35,
                    48,
                    49,
                ],
            },
            VcpCode {
                code: 214,
                values: [
                    1,
                    5,
                ],
            },
            VcpCode {
                code: 220,
                values: [
                    1,
                    2,
                    3,
                    4,
                    5,
                    6,
                    7,
                    8,
                ],
            },
            VcpCode {
                code: 223,
                values: [],
            },
            VcpCode {
                code: 224,
                values: [
                    0,
                    1,
                    2,
                    3,
                    4,
                    5,
                ],
            },
            VcpCode {
                code: 225,
                values: [
                    0,
                    1,
                ],
            },
            VcpCode {
                code: 227,
                values: [
                    0,
                    1,
                    2,
                    3,
                    4,
                    5,
                    6,
                ],
            },
            VcpCode {
                code: 228,
                values: [
                    0,
                    1,
                    2,
                    3,
                    4,
                    5,
                ],
            },
            VcpCode {
                code: 229,
                values: [
                    0,
                    1,
                    2,
                    3,
                ],
            },
            VcpCode {
                code: 230,
                values: [
                    0,
                    1,
                    2,
                    3,
                    4,
                ],
            },
            VcpCode {
                code: 231,
                values: [
                    0,
                    1,
                ],
            },
            VcpCode {
                code: 233,
                values: [
                    0,
                    1,
                ],
            },
            VcpCode {
                code: 234,
                values: [
                    0,
                    1,
                ],
            },
            VcpCode {
                code: 235,
                values: [
                    0,
                    1,
                ],
            },
        ],
    ),
}
